src/lndk_offers.rs

```diff
@@ -1,7 +1,189 @@
+use crate::blocking_client::BlockingClient;
+use bitcoin::network::constants::Network;
+use bitcoin::secp256k1::schnorr::Signature;
+use bitcoin::secp256k1::{Error as Secp256k1Error, Message, PublicKey};
+use lightning::offers::invoice_request::InvoiceRequest;
 use lightning::offers::offer::Offer;
 use lightning::offers::parse::Bolt12ParseError;
+use tonic_lnd::signrpc::{KeyLocator, SignMessageReq};
+use tonic_lnd::tonic::Status;
 
 // Decodes a bech32 string into an LDK offer.
 pub fn decode(offer_str: String) -> Result<Offer, Bolt12ParseError> {
     offer_str.parse::<Offer>()
 }
+
+#[allow(dead_code)]
+pub(crate) fn request_invoice(
+    mut signer: impl MessageSigner,
+    offer: Offer,
+    metadata: Vec<u8>,
+    network: Network,
+    msats: u64,
+) -> InvoiceRequest {
+    // Note that we need to use a blocking version of the LND client here because we need to use it in a
+    // blocking closure below to pass into LDK.
+    let key_loc = KeyLocator {
+        key_family: 6,
+        key_index: 1,
+    };
+
+    let pubkey_bytes = signer
+        .derive_key(key_loc.clone())
+        .expect("failed to derive key");
+    let pubkey = PublicKey::from_slice(&pubkey_bytes).expect("failed to deserialize public key");
+
+    let unsigned_invoice_req = match offer
+        .request_invoice(metadata, pubkey)
+        .unwrap()
+        .chain(network)
+        .unwrap()
+        .amount_msats(msats)
+        .unwrap()
+        .build()
+    {
+        Ok(req) => req,
+        Err(e) => panic!("ERROR requesting invoice request: {:?}", e),
+    };
+
+    // To create a valid invoice request, we also need to sign it.
+    let sign_closure = |msg: &Message| {
+        let signature_resp = signer
+            .sign_message(key_loc, msg.to_owned())
+            .map_err(|_| Secp256k1Error::InvalidSignature)?;
```

Comment:
Could this just be `block_on(asynchronous sign message call)`?  Then we wouldn't need a blocking/synchronous client at all?

Not sure if that'll work, async x closures often scream at me in funny ways but I think it would be nice if it works.

Reply:
I did actually start with the `block_on` option but had some trouble for some reason, I'll give it another whirl to remember what the error is. But in general I'm not a big fan of block_on because often it'll just hang infinitely without throwing an error and it's hard to gauge what's going on

Reply:
Ah so this seems like sort of the issue I ran into. If I try to use the normal tonic lnd client, and use `futures::executor::block_on(asynchronous sign message call)`, it hangs forever. Seems to be a deadlock and creating a new tokio runtime for the client is what's used instead. https://github.com/hyperium/tonic/issues/575#issuecomment-798760286

Another resource, which I used for the blocking tonic client: https://github.com/hyperium/tonic/blob/master/examples/src/blocking/client.rs

Reply:
Ok! Did some more research into this and looks like we can use `task::spawn_blocking` to avoid the BlockingClient :) I'll rework in that direction

Reply:
You're gonna be an async master after all of this 🙇

Reply:
Haha yeah definitely learning a lot!